bot/log.py

```diff
@@ -63,29 +38,50 @@ def setup_sentry() -> None:
         enable_tracing=True,
     )
 
+def _setup_file_logging(logger: logging.Logger) -> None:
+    """Set up file logging if enabled."""
+    if constants.FILE_LOGS:
+        log_file = Path("logs", "bot.log")
+        log_file.parent.mkdir(exist_ok=True)
+        file_handler = handlers.RotatingFileHandler(
+            log_file, maxBytes=5242880, backupCount=7, encoding="utf8"
+        )
+        file_handler.setFormatter(core_logging.log_format)
+        logger.addHandler(file_handler)
+
+def _setup_colored_logging(logger: logging.Logger) -> None:
+    """Set up colored logging output."""
+    if "COLOREDLOGS_LEVEL_STYLES" not in os.environ:
+        coloredlogs.DEFAULT_LEVEL_STYLES = {
+            **coloredlogs.DEFAULT_LEVEL_STYLES,
+            "trace": {"color": 246},
+            "critical": {"background": "red"},
+            "debug": coloredlogs.DEFAULT_LEVEL_STYLES["info"]
+        }
+    if "COLOREDLOGS_LOG_FORMAT" not in os.environ:
+        coloredlogs.DEFAULT_LOG_FORMAT = core_logging.log_format._fmt
+    coloredlogs.install(level=core_logging.TRACE_LEVEL, logger=logger, stream=sys.stdout)
 
 def _set_trace_loggers() -> None:
     """
     Set loggers to the trace level according to the value from the BOT_TRACE_LOGGERS env var.
 
     When the env var is a list of logger names delimited by a comma,
     each of the listed loggers will be set to the trace level.
-
     If this list is prefixed with a "!", all of the loggers except the listed ones will be set to the trace level.
-
     Otherwise if the env var begins with a "*",
     the root logger is set to the trace level and other contents are ignored.
     """
     level_filter = constants.Bot.trace_loggers
-    if level_filter:
-        if level_filter.startswith("*"):
-            get_logger().setLevel(core_logging.TRACE_LEVEL)
-
-        elif level_filter.startswith("!"):
-            get_logger().setLevel(core_logging.TRACE_LEVEL)
-            for logger_name in level_filter.strip("!,").split(","):
-                get_logger(logger_name).setLevel(logging.DEBUG)
-
-        else:
-            for logger_name in level_filter.strip(",").split(","):
-                get_logger(logger_name).setLevel(core_logging.TRACE_LEVEL)
+    if not level_filter:
+        return
+
+    if level_filter.startswith("*"):
+        get_logger().setLevel(core_logging.TRACE_LEVEL)
+    elif level_filter.startswith("!"):
+        get_logger().setLevel(core_logging.TRACE_LEVEL)
+        for logger_name in level_filter.strip("!,").split(","):
+            get_logger(logger_name).setLevel(logging.DEBUG)
+    else:
+        for logger_name in level_filter.strip(",").split(","):
+            get_logger(logger_name).setLevel(core_logging.TRACE_LEVEL)
```

Comment:
I thought ruff would complain once you have execution paths that do not return none when you have at least one that does. surprising.

Reply:
All paths in that function return None, just implicitly (one from a bare return statement, one from having no explicit return statement)

Reply:
> All paths in that function return None, just implicitly 

Yes that i know, it's just that i thought ruff would either impose that all returns be implicit, and when you have at leat one explicit, the rest need to become as such as well